internal/storage/metastorage/firestore/event_storage.go

```diff
@@ -42,47 +53,413 @@ func newEventStorage(params Params, client *firestore.Client) (internal.EventSto
-	panic("unimplemented")
+func (e *eventStorageImpl) GetMaxEventId(ctx context.Context, eventTag uint32) (int64, error) {
+	if err := e.validateEventTag(eventTag); err != nil {
+		return 0, err
+	}
+	return e.instrumentGetMaxEventId.Instrument(ctx, func(ctx context.Context) (int64, error) {
+		latestEvent, err := e.getLatestEventEntry(ctx, eventTag, nil)
+		if err != nil {
+			return 0, err
+		}
+		if latestEvent == nil {
+			return 0, errors.ErrNoEventHistory
+		}
+		return latestEvent.EventId, nil
+	})
 }
 
 // SetMaxEventId implements internal.EventStorage.
-func (*eventStorageImpl) SetMaxEventId(ctx context.Context, eventTag uint32, maxEventId int64) error {
-	panic("unimplemented")
+func (e *eventStorageImpl) SetMaxEventId(ctx context.Context, eventTag uint32, maxEventId int64) error {
+	if err := e.validateEventTag(eventTag); err != nil {
+		return err
+	}
+	if maxEventId < model.EventIdStartValue && maxEventId != model.EventIdDeleted {
+		return xerrors.Errorf("invalid max event id: %d", maxEventId)
+	}
+	return e.instrumentSetMaxEventId.Instrument(ctx, func(ctx context.Context) error {
+		return e.client.RunTransaction(ctx, func(ctx context.Context, t *firestore.Transaction) error {
+			latestEvent, err := e.getLatestEventEntry(ctx, eventTag, t)
+			if err != nil {
+				return err
+			}
+			if latestEvent != nil && latestEvent.EventId < maxEventId {
+				return xerrors.Errorf("can not set max event id to be %d, which is bigger than current max event id: %d", maxEventId, latestEvent.EventId)
+			}
+			if maxEventId != model.EventIdDeleted {
+				newLatestEventDocRef := e.getEventDocRef(eventTag, maxEventId)
+				newLatestEventDoc, err := t.Get(newLatestEventDocRef)
+				if err != nil {
+					return xerrors.Errorf("failed to get event entry with new max event id: %d, error: %w", maxEventId, err)
+				}
+				if !newLatestEventDoc.Exists() {
+					return xerrors.Errorf("event entry with new max event id %d does not exist", maxEventId)
+				}
+				err = t.Set(newLatestEventDocRef, map[string]bool{
+					"Latest": true,
+				}, firestore.MergeAll)
+				if err != nil {
+					return xerrors.Errorf("failed to set new max event id, error: %w", err)
+				}
+			}
+			if latestEvent != nil {
+				err = t.Set(e.getEventDocRef(eventTag, latestEvent.EventId), map[string]bool{
+					"Latest": false,
+				}, firestore.MergeAll)
+				if err != nil {
+					return xerrors.Errorf("failed to update previous max event id, error: %w", err)
+				}
+			}
+			return nil
+		})
+	})
+}
+
+/////////////////////////////////////////////////////////
+// firestore storage
+
+type firestoreEventEntryPayload struct {
+	EventId        int64
+	EventType      api.BlockchainEvent_Type
+	BlockHeight    int64
+	BlockHash      string
+	Tag            uint32
+	ParentHash     string
+	MaxEventId     int64
+	BlockSkipped   bool
+	EventTag       uint32
+	BlockTimestamp int64
+}
+
+type firestoreEventEntry struct {
+	EventTag    uint32
```

Comment:
Since EventTag is in every function, should eventTag be the collection instead of a field?

Reply:
Good idea, that could improve the performance.

Reply:
Done.